hs_core/models.py

```diff
@@ -1349,14 +1363,29 @@ class Meta:
         unique_together = ("content_type", "object_id")
 
 def get_path(instance, filename):
-    return os.path.join(instance.content_object.short_id, 'data', 'contents', filename)
+    if instance.content_object.resource_federation_path:
+        return os.path.join(instance.content_object.resource_federation_path, instance.content_object.short_id, 'data',
+                            'contents', filename)
+    else:
+        return os.path.join(instance.content_object.short_id, 'data', 'contents', filename)
 
 class ResourceFile(models.Model):
     object_id = models.PositiveIntegerField()
     content_type = models.ForeignKey(ContentType)
 
     content_object = GenericForeignKey('content_type', 'object_id')
-    resource_file = models.FileField(upload_to=get_path, max_length=500, storage=IrodsStorage() if getattr(settings,'USE_IRODS', False) else DefaultStorage())
+    resource_file = models.FileField(upload_to=get_path, max_length=500, null=True, blank=True,
+                                     storage=IrodsStorage() if getattr(settings,'USE_IRODS', False) else DefaultStorage())
+    # the following optional fields are added for use by federated iRODS resources where resources are created in the local federated
+    # zone rather than hydroshare zone, in which case resource_file is empty, and we record iRODS logical resource file name and
+    # file size for customized copy or move operations for resource creation and other operations
+    # fed_resource_file in particular is a counterpart of resource_file, but handles files uploaded from local disk and store the files
+    # to federated zone rather than hydroshare zone.
+    fed_resource_file = models.FileField(upload_to=get_path, max_length=500, null=True, blank=True,
+                                     storage=IrodsStorage('federated') if getattr(settings,'USE_IRODS', False) else DefaultStorage())
+    fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
```

Comment:
@pkdash ```fed_resource_file``` is for handling files uploaded from local disk, but instead of storing it to hydroshare zone as we are currently doing (i.e., where resource_file is being used), the uploaded file is stored to a federated zone. ```fed_resource_file_name_or_path``` is for handling and recording resource file names browsed from a federated zone via iRODS browser, in which case, we cannot use a FileField since the file is not uploaded from local disk, and a string representing iRODS collection or data object logical path is sufficient to locate file directly from a federated zone. Hope this helps answer your questions.